### PR TITLE
Respect SchedulingPolicy

### PR DESCRIPTION
**Pull request by tenzen-y**
Signed-off-by: Yuki Iwai <yuki.iwai.tz@gmail.com>

I changed the logic for the gang-scheduling so that the mpi-operator respects SchedulingPolicy when creating PodGroup.

Fixes: #518 

/assign @alculquicondor

### CODE REVIEW

**Comment by alculquicondor** [comment]
Can you summarize the changes for somebody that isn't familiar with volcano? :sweat_smile:

**Comment by tenzen-y** [comment]
> Can you summarize the changes for somebody that isn't familiar with volcano? 😅

Sure. I will send a ping to you once this PR description is ready.

File: pkg/controller/podgroup.go [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+	}
+	if l := replicas[kubeflow.MPIReplicaTypeLauncher]; l != nil && len(l.Template.Spec.PriorityClassName) != 0 {
+		return l.Template.Spec.PriorityClassName
+	} else if w := replicas[kubeflow.MPIReplicaTypeWorker]; w != nil && len(w.Template.Spec.PriorityClassName) != 0 {
+		return w.Template.Spec.PriorityClassName
+	}
+	return ""
+}
+
+func (c *MPIJobController) calcPGMinResources(
```
Comment (alculquicondor) [inline pkg/controller/podgroup.go:89]: add a comment for what's happening here

File: pkg/controller/mpi_job_controller.go [hunk 2]
```diff
@@ -388,13 +401,20 @@ func (c *MPIJobController) Run(threadiness int, stopCh <-chan struct{}) error {
...
+		c.jobSynced,
+		c.podSynced,
+		c.priorityClassSynced,
+		c.mpiJobSynced,
 	}
 	if c.gangSchedulerName != "" {
-		if ok := cache.WaitForCacheSync(stopCh, c.podgroupsSynced); !ok {
-			return fmt.Errorf("failed to wait for podgroup caches to sync")
-		}
+		synced = append(synced, c.podgroupsSynced)
```
[the remark this answers is not in the document]
Comment (tenzen-y) [reply pkg/controller/mpi_job_controller.go:414]: Yes, that's right. 
It would be good to sync priorityclass only when the gangSchedulerName is set, same as priorityclass.

[the remark this answers is not in the document]
Comment (tenzen-y) [reply pkg/controller/mpi_job_controller.go:414]: I will change the above logic.

File: pkg/controller/podgroup.go [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+		}
+		if len(schedulingPolicy.Queue) != 0 {
+			queueName = schedulingPolicy.Queue
+		}
+		if schedulingPolicy.MinResources != nil {
+			minResources = schedulingPolicy.MinResources
+		}
+	}
+	if minResources == nil {
+		minResources = c.calcPGMinResources(minMember, mpiJob.Spec.MPIReplicaSpecs)
```
[the remark this answers is not in the document]
Comment (tenzen-y) [reply pkg/controller/podgroup.go:55]: Yes. Previously, we could not use `minResources` since we don't calculate and set that:

https://github.com/kubeflow/mpi-operator/blob/c21942d1e27d196d06078f67f39a6f7a0e990dee/pkg/controller/mpi_job_controller.go#L1299-L1324

[the remark this answers is not in the document]
Comment (tenzen-y) [reply pkg/controller/podgroup.go:55]: If we configurable the `minResources`, scheduling part of Workers to Nodes is useful when the number of Workers is huge.

File: pkg/controller/podgroup.go [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+// newPodGroup creates a new PodGroup for an MPIJob
+// resource. It also sets the appropriate OwnerReferences on the resource so
+// handleObject can discover the MPIJob resource that 'owns' it.
+func (c *MPIJobController) newPodGroup(mpiJob *kubeflow.MPIJob) *volcanov1beta1.PodGroup {
+	minMember := workerReplicas(mpiJob) + 1
+	queueName := mpiJob.Annotations[volcanov1beta1.QueueNameAnnotationKey]
+	var minResources *corev1.ResourceList
+	if schedulingPolicy := mpiJob.Spec.RunPolicy.SchedulingPolicy; schedulingPolicy != nil {
+		if schedulingPolicy.MinAvailable != nil {
+			minMember = *schedulingPolicy.MinAvailable
```
[the remark this answers is not in the document]
Comment (tenzen-y) [reply pkg/controller/podgroup.go:45]: > what if volcano schedules a number of pods that is less than workerReplicas(mpiJob)?
The mpirun command might get stuck waiting for all the workers (unless it's elastic horovod that has native support for changing number of workers).

In that case, volcano schedules all Pods. If the cluster does not have enough resources for all Pods, some pods will be marked as `Pending`.

> We need a way to restrict the usage of this field or have clear documentation that this is not supported for all kinds of MPI applications.

I agree. I think we have 2 options. Since we don't have fields if the MPIJob uses elastic horovod. 

1. As you say, have clear documentation.
2. Add a new member, 'Elastic bool `json:elastic,omitempty`' into `Runpolicy` to represent if MPIJob uses elastic mode. And then, we add validations for `minMember` and `minResources`.

I like the second option. What do you think? @alculquicondor

**Comment by tenzen-y** [comment]
@alculquicondor Updated PR description.

File: pkg/controller/podgroup.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (alculquicondor) [reply pkg/controller/podgroup.go:45]: Are there similar fields in the training operator?

File: pkg/controller/mpi_job_controller.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (alculquicondor) [reply pkg/controller/mpi_job_controller.go:414]: still not changed

File: pkg/controller/podgroup.go [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+
+type replicaPriority struct {
+	priority int32
+
+	common.ReplicaSpec
+}
+
+// newPodGroup creates a new PodGroup for an MPIJob
+// resource. It also sets the appropriate OwnerReferences on the resource so
+// handleObject can discover the MPIJob resource that 'owns' it.
```
Comment (alculquicondor) [inline pkg/controller/podgroup.go:38]: explain how minResources is calculated.

What does volcano use that field? If it's all added up, how does it know whether they are consumed by one or multiple pods?

File: pkg/controller/podgroup.go (hunk 1 above)
Comment (alculquicondor) [reply pkg/controller/podgroup.go:89]: Still waiting.

What does volcano use that field? If it's all added up, how does it know whether they are consumed by one or multiple pods?

File: pkg/controller/podgroup.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (tenzen-y) [reply pkg/controller/podgroup.go:45]: No, most customjob-controllers have no fields on whether CustomJob uses elastic mode. This means users can set any value to `schedulingPolicy.MinMember`. The pytorchjob-controller has `ElasticPolicy`, although that field doesn't affect schedulingPolicy.

https://github.com/kubeflow/training-operator/blob/6ce98387a1a68b79365495a6344f7f19b48557d3/pkg/apis/kubeflow.org/v1/pytorch_types.go#L79

[the remark this answers is not in the document]
Comment (alculquicondor) [reply pkg/controller/podgroup.go:45]: in that case, I think we should proceed with option 1 for now.

File: pkg/controller/podgroup.go (hunk 5 above)
Comment (tenzen-y) [reply pkg/controller/podgroup.go:38]: > explain how minResources is calculated.

Sure.

> What does volcano use that field? If it's all added up, how does it know whether they are consumed by one or multiple pods?

I will answer in another thread.

File: pkg/controller/podgroup.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (tenzen-y) [reply pkg/controller/podgroup.go:45]: > in that case, I think we should proceed with option 1 for now.

I agree. Maybe, we can add documentation to https://www.kubeflow.org/docs/components/training/job-scheduling/.

> That said, there is currently no documentation about SchedulingPolicy overall 

You're right...

File: pkg/controller/podgroup.go (hunk 1 above)
Comment (tenzen-y) [reply pkg/controller/podgroup.go:89]: > What does volcano use that field? If it's all added up, how does it know whether they are consumed by one or multiple pods?

I tried to dive into the volcano. The volcano has the Queue resource like ClusterQueue in Kueue. 
If Queue doesn't have enough resources requested by `minResources`  in PodGroup, volcano-scheduler doesn't schedule all Pods to Nodes.

Does that make sense?

Comment (alculquicondor) [reply pkg/controller/podgroup.go:89]: makes sense.... it looks like a quota system just like kueue. I wonder if they would be interested in reviewing this PR. Do you know anybody?

Comment (tenzen-y) [reply pkg/controller/podgroup.go:89]: I remember @zw0610 is familiar with volcano. And @zw0610 is a reviewer in this repository :)

Comment (tenzen-y) [reply pkg/controller/podgroup.go:89]: IMO, as a result of investigating the volcano, maybe we don't need the `calcPGMinResources` function since `minResources` don't affect gang-scheduling. And we may be able to propagate `schedulingPolicy.minResources` to PodGroup.

Comment (tenzen-y) [reply pkg/controller/podgroup.go:89]: @zw0610 Let us know what you think.

Comment (tenzen-y) [reply pkg/controller/podgroup.go:89]: Maybe, @shinytang6 is also interested in this PR.
cc: @shinytang6

Comment (tenzen-y) [reply pkg/controller/podgroup.go:89]: @alculquicondor They don't seem to have enough bandwidth to review this PR.
So, I'm thinking of removing `func calcPGMinResources` and then passing `SchedulingPolicy.MinResources` to `PodGroup.Spec.MinResources`.

Also, it would be good to keep tracking this.

Comment (alculquicondor) [reply pkg/controller/podgroup.go:89]: SGTM, we don't want to introduce functionality we aren't sure about.

Comment (tenzen-y) [reply pkg/controller/podgroup.go:89]: > we don't want to introduce functionality we aren't sure about.

I agree.

File: pkg/controller/podgroup.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (tenzen-y) [reply pkg/controller/podgroup.go:45]: > So we need to clarify it in the documentation anyways.

I agree.

Probably, we can discuss if we add an `elastic` field in other places (issue or pr).
In this PR, we can just pass `SchedulingPolicy.MinAvailable` to `PodGroup.Spec.MinMember` and then add docs about that.

File: pkg/controller/mpi_job_controller.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (tenzen-y) [reply pkg/controller/mpi_job_controller.go]: Done.

File: pkg/controller/podgroup.go (hunk 1 above)
Comment (tenzen-y) [reply pkg/controller/podgroup.go]: Done.

**Comment by tenzen-y** [comment]
@alculquicondor I have addressed your comments. Please take another look.

File: cmd/mpi-operator/app/server.go [hunk 6]
```diff
@@ -46,6 +46,7 @@ import (
 
 	"github.com/kubeflow/mpi-operator/cmd/mpi-operator/app/options"
 	mpijobclientset "github.com/kubeflow/mpi-operator/pkg/client/clientset/versioned"
+	kubeflowScheme "github.com/kubeflow/mpi-operator/pkg/client/clientset/versioned/scheme"
```
Comment (alculquicondor) [inline cmd/mpi-operator/app/server.go:49]:
```suggestion
	kubeflowscheme "github.com/kubeflow/mpi-operator/pkg/client/clientset/versioned/scheme"
```

File: pkg/apis/kubeflow/v2beta1/types.go [hunk 7]
```diff
@@ -54,11 +54,35 @@ const (
...
 // job, for example `minAvailable` for gang-scheduling.
 type SchedulingPolicy struct {
-	MinAvailable           *int32           `json:"minAvailable,omitempty"`
-	Queue                  string           `json:"queue,omitempty"`
-	MinResources           *v1.ResourceList `json:"minResources,omitempty"`
-	PriorityClass          string           `json:"priorityClass,omitempty"`
-	ScheduleTimeoutSeconds *int32           `json:"scheduleTimeoutSeconds,omitempty"`
+	// MinAvailable defines the minimal number of member to run the PodGroup.
+	// If the gang-scheduling is set to the volcano,
+	// input is passed to `.spec.mimMember` in PodGroup for the volcano.
```
Comment (alculquicondor) [inline pkg/apis/kubeflow/v2beta1/types.go:59]: Explain that this will only work if the MPI application supports resizing.

File: pkg/controller/mpi_job_controller.go [hunk 8]
```diff
@@ -364,6 +376,13 @@ func NewMPIJobControllerWithClock(
 		UpdateFunc: controller.handleObjectUpdate,
 		DeleteFunc: controller.handleObject,
 	})
+	if priorityClassInformer != nil {
+		priorityClassInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+			AddFunc:    controller.handleObject,
+			UpdateFunc: controller.handleObjectUpdate,
+			DeleteFunc: controller.handleObject,
```
Comment (alculquicondor) [inline pkg/controller/mpi_job_controller.go:383]: Does it make sense to add an event handler for priorityClass? They are not owned by the MPIJob.

File: pkg/controller/podgroup.go (hunk 5 above)
Comment (alculquicondor) [reply pkg/controller/podgroup.go:28]: add in the comment the fields that this function populates and how

File: pkg/controller/podgroup.go [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+//  2. .spec.mpiReplicaSecs[Launcher].template.spec.priorityClassName
+//  3. .spec.mpiReplicaSecs[Worker].template.spec.priorityClassName
+//
+// However, if the calculated priorityClass name is missing in the cluster, return an empty.
+func (c *MPIJobController) calcPriorityClassName(
+	replicas map[kubeflow.MPIReplicaType]*common.ReplicaSpec,
+	schedulingPolicy *kubeflow.SchedulingPolicy,
+) string {
+	if schedulingPolicy != nil && len(schedulingPolicy.PriorityClass) != 0 &&
+		c.priorityClassExist(schedulingPolicy.PriorityClass) {
```
Comment (alculquicondor) [inline pkg/controller/podgroup.go:72]: Should MPIJob care if it exists? I think it's up to Volcano to decide what to do.

File: pkg/apis/kubeflow/v2beta1/types.go (hunk 7 above)
Comment (tenzen-y) [reply pkg/apis/kubeflow/v2beta1/types.go:59]: Sounds good.

File: pkg/controller/mpi_job_controller.go (hunk 8 above)
Comment (tenzen-y) [reply pkg/controller/mpi_job_controller.go:383]: You're right.

File: pkg/controller/podgroup.go (hunk 9 above)
Comment (tenzen-y) [reply pkg/controller/podgroup.go:72]: Makes sense.

File: cmd/mpi-operator/app/server.go (hunk 6 above)
Comment (tenzen-y) [reply cmd/mpi-operator/app/server.go]: Done.

File: pkg/apis/kubeflow/v2beta1/types.go (hunk 7 above)
Comment (tenzen-y) [reply pkg/apis/kubeflow/v2beta1/types.go:59]: Done.

File: pkg/controller/podgroup.go (hunk 9 above)
Comment (tenzen-y) [reply pkg/controller/podgroup.go:72]: We can leave it to the volcano.

Comment (alculquicondor) [reply pkg/controller/podgroup.go:72]: Then let's cleanup the informer

Comment (tenzen-y) [reply pkg/controller/podgroup.go]: Done.

File: pkg/controller/mpi_job_controller.go (hunk 8 above)
Comment (tenzen-y) [reply pkg/controller/mpi_job_controller.go]: Done.

File: pkg/controller/podgroup.go (hunk 5 above)
Comment (tenzen-y) [reply pkg/controller/podgroup.go:36]: Done.

**Comment by tenzen-y** [comment]
@alculquicondor Updated. PTAL.

File: pkg/apis/kubeflow/v2beta1/types.go [hunk 10]
```diff
@@ -54,11 +54,36 @@ const (
...
+	// Also, this parameter will function properly only when we use elastic training (e.g., Elastic Horovod).
```
Comment (alculquicondor) [inline pkg/apis/kubeflow/v2beta1/types.go:60]:
```suggestion
	// When using this field, you need to make sure the application supports resizing (e.g., Elastic Horovod).
```

File: pkg/apis/kubeflow/v2beta1/types.go (hunk 7 above)
Comment (alculquicondor) [reply pkg/apis/kubeflow/v2beta1/types.go:59]: If not set, it defaults to the number of workers

**Comment by tenzen-y** [comment]
@alculquicondor I addressed your suggestions and squashed commits into one.

**Comment by tenzen-y** [comment]
Also, I will add docs for the schedulingPolicy to https://www.kubeflow.org/.

**Comment by tenzen-y** [comment]
@alculquicondor Created https://github.com/kubeflow/website/pull/3453.

**Comment by alculquicondor** [comment]
oops, no, I just forgot to approve
/approve
